File: exercises/python/order_service/infrastructure/payment_adapter.py
"""Thin REST client adapter for PaymentPort.

No full Anti-Corruption Layer here because Order and Payment share
vocabulary closely enough that the cost of a full ACL outweighs the
benefit. We do still:

- Name the span in domain language (``Order.Payment.Authorize``)
- Tag the span with business attributes for trace search
- Convert transport failures to typed outcomes so the saga's pattern
  matching remains exhaustive
"""

from __future__ import annotations

import os
from typing import Any

import httpx
import structlog
from opentelemetry import trace
from opentelemetry.propagate import inject

from domain.models import Order
from domain.ports import (
    AuthorizationFailure,
    AuthorizationOutcome,
    Authorized,
    Declined,
)

logger = structlog.get_logger()
tracer = trace.get_tracer("order-service")

_DEFAULT_PAYMENT_URL = "http://payment-service:8082"


class PaymentRestAdapter:
    """REST client adapter implementing PaymentPort.

    Thin client -- no separate DTO package, no drift counter. The
    vocabularies are close enough that translation cost outweighs
    translation benefit.
    """

    def __init__(self, base_url: str | None = None) -> None:
        self._base_url = (
            base_url
            or os.environ.get("PAYMENT_SERVICE_URL", _DEFAULT_PAYMENT_URL)
        )
        self._client = httpx.Client(timeout=30.0)

    def authorize(self, order: Order) -> AuthorizationOutcome:
        with tracer.start_as_current_span("Order.Payment.Authorize") as span:
            payment_method = "credit_card"
            span.set_attribute("payment.method", payment_method)
            span.set_attribute("order.value", float(order.total().amount))

            try:
                wire_request: dict[str, Any] = {
                    "orderId": order.id.value,
                    "customerId": order.customer_id.value,
                    "amount": float(order.total().amount),
                    "currency": order.total().currency,
                    "paymentMethod": payment_method,
                }

                # Inject OTel trace context for distributed tracing
                headers: dict[str, str] = {}
                inject(headers)

                url = f"{self._base_url}/api/payments/authorize"
                response = self._client.post(
                    url, json=wire_request, headers=headers
                )
                response.raise_for_status()
                wire_response = response.json()

                outcome = wire_response.get("outcome")

                if outcome == "AUTHORIZED":
                    auth_id = wire_response.get("authorizationId")
                    return Authorized(authorization_id=auth_id)

                if outcome == "DECLINED":
                    reason = wire_response.get("reason") or "declined"
                    return Declined(reason=reason)

                # FAILURE or unknown outcome
                reason = wire_response.get("reason") or f"unknown outcome: {outcome}"
                return AuthorizationFailure(detail=reason)

            except httpx.HTTPStatusError as e:
                logger.warning(
                    "Payment REST call failed",
                    status=e.response.status_code,
                    error=str(e),
                )
                return AuthorizationFailure(
                    detail=f"Payment REST call failed: HTTP {e.response.status_code}"
                )

            except httpx.HTTPError as e:
                logger.warning(
                    "Payment REST call failed", error=str(e)
                )
                return AuthorizationFailure(
                    detail=f"Payment REST call failed: {e}"
                )

File: exercises/dotnet/OrderService/Infrastructure/PaymentAdapter.cs
using System.Diagnostics;
using System.Net.Http.Json;
using System.Text.Json;
using OrderService.Domain;

namespace OrderService.Infrastructure;

public sealed class PaymentRestAdapter : IPaymentPort
{
    private static readonly ActivitySource ActivitySource = new("order-service");
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    private readonly HttpClient _httpClient;

    public PaymentRestAdapter(HttpClient httpClient)
    {
        _httpClient = httpClient;
    }

    public async Task<AuthorizationOutcome> Authorize(Order order)
    {
        var total = order.Total();

        using var activity = ActivitySource.StartActivity("Order.Payment.Authorize");
        activity?.SetTag("payment.method", "credit_card");
        activity?.SetTag("order.value", (double)total.Amount);

        try
        {
            var wireRequest = new
            {
                orderId = order.Id.Value,
                customerId = order.CustomerId.Value,
                amount = total.Amount,
                currency = total.Currency,
                paymentMethod = "credit_card"
            };

            var response = await _httpClient.PostAsJsonAsync(
                "/api/payments/authorize", wireRequest, JsonOptions);

            response.EnsureSuccessStatusCode();

            var responseBody = await response.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(responseBody);
            var root = doc.RootElement;
            var outcome = root.GetProperty("outcome").GetString();

            return outcome switch
            {
                "AUTHORIZED" => new AuthorizationOutcome.Authorized(
                    root.GetProperty("authorizationId").GetString()
                        ?? throw new InvalidOperationException("Missing authorizationId in AUTHORIZED response")),
                "DECLINED" => new AuthorizationOutcome.Declined(
                    root.TryGetProperty("reason", out var dr) ? dr.GetString() ?? "Declined" : "Declined"),
                _ => new AuthorizationOutcome.AuthorizationFailure(
                    $"Unknown payment outcome: {outcome}")
            };
        }
        catch (HttpRequestException ex)
        {
            return new AuthorizationOutcome.AuthorizationFailure(
                $"Payment service error: {ex.Message}");
        }
    }
}
